File: programming-language/Test/numerical_methods/question20.R
# Método de Newton modificado para aceitar um intervalo para a raiz
newton_method_interval <- function(y, interval, epsilon = 0.0001, max_iter = 1000) {
  a <- interval[1]
  b <- interval[2]
  mu <- (a + b) / 2  # Ponto inicial como o ponto médio do intervalo
  
  for (i in 1:max_iter) {
    mu_new <- mu - f(mu, y) / f_prime(mu, y)
    if (abs(mu_new - mu) < epsilon) {
      return(mu_new)
    }
    mu <- mu_new
    
    # Verifica se a raiz está dentro do intervalo
    if (mu < a || mu > b) {
      stop("O método de Newton saiu do intervalo de busca.")
    }
  }
  stop("O método de Newton não convergiu após ", max_iter, " iterações.")
}

# Valores de y
y <- c(4, 6, 5, 7, 7, 2, 5, 7, 5, 5)

# Escolhendo um intervalo para a raiz
interval <- c(-2, 3)

# Chamando o método de Newton modificado
raiz <- newton_method_interval(y, interval)

# Imprimindo o resultado
cat("A raiz à direita do ponto crítico é:", round(raiz, 3))

File: programming-language/Test/numerical_methods/question20.py
import numpy as np

# Função f e sua derivada em relação a μ
def f(mu, y):
    n = len(y)
    result = 0
    for yi in y:
        term1 = yi * np.log(yi / 10 / mu)
        term2 = (10 - yi) * np.log((10 - yi) / 10 / (1 - mu))
        result += 2 * (term1 + term2)
    return result - 3.84

def df(mu, y):
    n = len(y)
    result = 0
    for yi in y:
        term1 = yi / mu - (10 - yi) / (1 - mu)
        term2 = yi / (mu ** 2) + (10 - yi) / ((1 - mu) ** 2)
        result += 2 * term1 * term2
    return result

# Método de Newton para encontrar a raiz
def newton_method(y, mu_guess, epsilon=0.0001, max_iterations=1000):
    mu = mu_guess
    for _ in range(max_iterations):
        delta_mu = f(mu, y) / df(mu, y)
        mu -= delta_mu
        if abs(delta_mu) < epsilon:
            return mu
    return mu

# Valores de y fornecidos
y = [4, 6, 5, 7, 7, 2, 5, 7, 5, 5]

# Chute inicial
mu_guess = 0.5

# Encontrar a raiz
raiz = newton_method(y, mu_guess)

print("A raiz a direita do ponto crítico é:", round(raiz, 3))